File: tests/test_class_sh_trampoline_shared_ptr_cpp_arg.py
import pytest

import pybind11_tests.class_sh_trampoline_shared_ptr_cpp_arg as m


def test_shared_ptr_cpp_arg():
    import weakref

    class PyChild(m.SpBase):
        def is_base_used(self):
            return False

    tester = m.SpBaseTester()

    obj = PyChild()
    objref = weakref.ref(obj)

    # Pass the last python reference to the C++ function
    tester.set_object(obj)
    del obj
    pytest.gc_collect()

    # python reference is still around since C++ has it now
    assert objref() is not None
    assert tester.is_base_used() is False
    assert tester.obj.is_base_used() is False
    assert tester.get_object() is objref()


def test_shared_ptr_cpp_prop():
    class PyChild(m.SpBase):
        def is_base_used(self):
            return False

    tester = m.SpBaseTester()

    # Set the last python reference as a property of the C++ object
    tester.obj = PyChild()
    pytest.gc_collect()

    # python reference is still around since C++ has it now
    assert tester.is_base_used() is False
    assert tester.has_python_instance() is True
    assert tester.obj.is_base_used() is False
    assert tester.obj.has_python_instance() is True


def test_shared_ptr_arg_identity():
    import weakref

    tester = m.SpBaseTester()

    obj = m.SpBase()
    objref = weakref.ref(obj)

    tester.set_object(obj)
    del obj
    pytest.gc_collect()

    # SMART_HOLDER_WIP: the behavior below is DIFFERENT from PR #2839
    # python reference is gone because it is not an Alias instance
    assert objref() is None
    assert tester.has_python_instance() is False


def test_shared_ptr_alias_nonpython():
    tester = m.SpBaseTester()

    # C++ creates the object, a python instance shouldn't exist
    tester.set_nonpython_instance()
    assert tester.is_base_used() is True
    assert tester.has_instance() is True
    assert tester.has_python_instance() is False

    # Now a python instance exists
    cobj = tester.get_object()
    assert cobj.has_python_instance()
    assert tester.has_instance() is True
    assert tester.has_python_instance() is True

    # Now it's gone
    del cobj
    pytest.gc_collect()
    assert tester.has_instance() is True
    assert tester.has_python_instance() is False

    # When we pass it as an arg to a new tester the python instance should
    # disappear because it wasn't created with an alias
    new_tester = m.SpBaseTester()

    cobj = tester.get_object()
    assert cobj.has_python_instance()

    new_tester.set_object(cobj)
    assert tester.has_python_instance() is True
    assert new_tester.has_python_instance() is True

    del cobj
    pytest.gc_collect()

    # Gone!
    assert tester.has_instance() is True
    assert tester.has_python_instance() is False
    assert new_tester.has_instance() is True
    assert new_tester.has_python_instance() is False


def test_shared_ptr_goaway():
    import weakref

    tester = m.SpGoAwayTester()

    obj = m.SpGoAway()
    objref = weakref.ref(obj)

    assert tester.obj is None

    tester.obj = obj
    del obj
    pytest.gc_collect()

    # python reference is no longer around
    assert objref() is None
    # C++ reference is still around
    assert tester.obj is not None


def test_infinite():
    tester = m.SpBaseTester()
    while True:
        tester.set_object(m.SpBase())
        break  # Comment out for manual leak checking (use `top` command).


@pytest.mark.parametrize(
    "pass_through_func", [m.pass_through_shd_ptr, m.pass_through_shd_ptr_release_gil]
)
def test_std_make_shared_factory(pass_through_func):
    class PyChild(m.SpBase):
        def __init__(self):
            super().__init__(0)

    obj = PyChild()
    while True:
        assert pass_through_func(obj) is obj
        break  # Comment out for manual leak checking (use `top` command).

File: tests/test_class_sh_trampoline_shared_ptr_cpp_arg.cpp
// Copyright (c) 2021 The Pybind Development Team.
// All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

#include "pybind11/smart_holder.h"
#include "pybind11_tests.h"

#include <utility>

namespace {

// For testing whether a python subclass of a C++ object dies when the
// last python reference is lost
struct SpBase {
    // returns true if the base virtual function is called
    virtual bool is_base_used() { return true; }

    // returns true if there's an associated python instance
    bool has_python_instance() {
        auto *tinfo = py::detail::get_type_info(typeid(SpBase));
        return (bool) py::detail::get_object_handle(this, tinfo);
    }

    SpBase() = default;
    SpBase(const SpBase &) = delete;
    virtual ~SpBase() = default;
};

std::shared_ptr<SpBase> pass_through_shd_ptr(const std::shared_ptr<SpBase> &obj) { return obj; }

struct PySpBase : SpBase {
    using SpBase::SpBase;
    bool is_base_used() override { PYBIND11_OVERRIDE(bool, SpBase, is_base_used); }
};

struct SpBaseTester {
    std::shared_ptr<SpBase> get_object() const { return m_obj; }
    void set_object(std::shared_ptr<SpBase> obj) { m_obj = std::move(obj); }
    bool is_base_used() { return m_obj->is_base_used(); }
    bool has_instance() { return (bool) m_obj; }
    bool has_python_instance() { return m_obj && m_obj->has_python_instance(); }
    void set_nonpython_instance() { m_obj = std::make_shared<SpBase>(); }
    std::shared_ptr<SpBase> m_obj;
};

// For testing that a C++ class without an alias does not retain the python
// portion of the object
struct SpGoAway {};

struct SpGoAwayTester {
    std::shared_ptr<SpGoAway> m_obj;
};

} // namespace

PYBIND11_SMART_HOLDER_TYPE_CASTERS(SpBase)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(SpBaseTester)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(SpGoAway)
PYBIND11_SMART_HOLDER_TYPE_CASTERS(SpGoAwayTester)

TEST_SUBMODULE(class_sh_trampoline_shared_ptr_cpp_arg, m) {
    // For testing whether a python subclass of a C++ object dies when the
    // last python reference is lost

    py::classh<SpBase, PySpBase>(m, "SpBase")
        .def(py::init<>())
        .def(py::init([](int) { return std::make_shared<PySpBase>(); }))
        .def("is_base_used", &SpBase::is_base_used)
        .def("has_python_instance", &SpBase::has_python_instance);

    m.def("pass_through_shd_ptr", pass_through_shd_ptr);
    m.def("pass_through_shd_ptr_release_gil",
          pass_through_shd_ptr,
          py::call_guard<py::gil_scoped_release>()); // PR #4196

    py::classh<SpBaseTester>(m, "SpBaseTester")
        .def(py::init<>())
        .def("get_object", &SpBaseTester::get_object)
        .def("set_object", &SpBaseTester::set_object)
        .def("is_base_used", &SpBaseTester::is_base_used)
        .def("has_instance", &SpBaseTester::has_instance)
        .def("has_python_instance", &SpBaseTester::has_python_instance)
        .def("set_nonpython_instance", &SpBaseTester::set_nonpython_instance)
        .def_readwrite("obj", &SpBaseTester::m_obj);

    // For testing that a C++ class without an alias does not retain the python
    // portion of the object

    py::classh<SpGoAway>(m, "SpGoAway").def(py::init<>());

    py::classh<SpGoAwayTester>(m, "SpGoAwayTester")
        .def(py::init<>())
        .def_readwrite("obj", &SpGoAwayTester::m_obj);
}
